### PR TITLE
Add segment reference lock for bulk load

### PR DESCRIPTION
**Pull request by soothing-rain**
So we can support DataQueryable and DataIndexed states

issue: #16607
Signed-off-by: Yuchen Gao <yuchen.gao@zilliz.com>

### CODE REVIEW

File: internal/rootcoord/root_coord.go [hunk 1]
```diff
@@ -2342,6 +2400,20 @@ func (c *Core) ReportImport(ctx context.Context, ir *rootcoordpb.ImportResult) (
...
 	}
+	// Special case for ImportState_ImportAllocSegment state, where we shall only add segment ref lock and do no other
+	// operations.
+	// TODO: This is inelegant and must get re-structured.
+	if ir.GetState() == commonpb.ImportState_ImportAllocSegment {
+		// Lock the segments, so we don't lose track of them when compaction happens.
+		// Note that these locks will be unlocked in c.postImportPersistLoop() -> checkSegmentLoadedLoop().
+		if err := c.CallAddSegRefLock(ctx, ir.GetSegments()); err != nil {
+			// Fail-open.
+			log.Error("failed to acquire segment ref lock", zap.Error(err))
```
Comment (xiaocai2333) [inline internal/rootcoord/root_coord.go:2411]: If the segment lock fails, can it be handled normally?

File: internal/rootcoord/root_coord.go [hunk 2]
```diff
@@ -2519,19 +2610,25 @@ func (c *Core) checkSegmentLoadedLoop(ctx context.Context, taskID int64, colID i
 	defer ticker.Stop()
 	expireTicker := time.NewTicker(time.Duration(Params.RootCoordCfg.ImportSegmentStateWaitLimit*1000) * time.Millisecond)
 	defer expireTicker.Stop()
+	defer c.CallReleaseSegRefLock(ctx, segIDs)
```
Comment (xiaocai2333) [inline internal/rootcoord/root_coord.go:2613]: If an error occurred between locking and releasing the lock, can the lock be released successfully?

File: internal/rootcoord/root_coord.go (hunk 1 above)
Comment (xiaocai2333) [reply internal/rootcoord/root_coord.go:2411]: I see no error is returned if something error here.

File: internal/rootcoord/root_coord.go (hunk 2 above)
Comment (soothing-rain) [reply internal/rootcoord/root_coord.go]: locks will now be released when an import task completes, fails, or expires.

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**